Index: src/map/skill.c
===================================================================
--- src/map/skill.c	(revision 16499)
+++ src/map/skill.c	(working copy)
@@ -9431,7 +9431,7 @@
 	case MO_BODYRELOCATION:
 		if (unit_movepos(src, x, y, 1, 1)) {
 			clif_skill_poseffect(src,skillid,skilllv,src->x,src->y,tick);
-//			clif_slide(src, src->x, src->y); //Poseffect is the one that makes the char snap on the client...
+			clif_slide(src, src->x, src->y); //Poseffect is the one that makes the char snap on the client...
 			if (sd) skill_blockpc_start (sd, MO_EXTREMITYFIST, 2000);
 		}
 		break;
